# **Book Recommendation System**

### Know the data

In [1]:
# Libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Load datasets
path_book= '/content/drive/MyDrive/MyProjects/ML Projects/Unsupervised ML/BookRecommendationSystem/data/Books.csv'
path_rating= '/content/drive/MyDrive/MyProjects/ML Projects/Unsupervised ML/BookRecommendationSystem/data/Ratings.csv'
path_users= '/content/drive/MyDrive/MyProjects/ML Projects/Unsupervised ML/BookRecommendationSystem/data/Users.csv'

book_data= pd.read_csv(path_book)
rating_data= pd.read_csv(path_rating)
users_data= pd.read_csv(path_users)
data_dict= {'book_data':book_data, 'rating_data':rating_data, 'users_data':users_data}

<ipython-input-2-15ef8e7ae510>:6: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  book_data= pd.read_csv(path_book)


In [3]:
# First view
display(book_data.head(3))
display(rating_data.head(3))
display(users_data.head(3))

ISBN            Book-Title           Book-Author Year-Of-Publication  \
0  0195153448   Classical Mythology    Mark P. O. Morford                2002   
1  0002005018          Clara Callan  Richard Bruce Wright                2001   
2  0060973129  Decision in Normandy          Carlo D'Este                1991   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
2          HarperPerennial  http://images.amazon.com/images/P/0060973129.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0

User-ID                         Location   Age
0        1               nyc, new york, usa   NaN
1        2        stockton, california, usa  18.0
2        3  moscow, yukon territory, russia   NaN

In [4]:
# Shape
for key in data_dict.keys():
  print(f'{key} shape : {data_dict[key].shape}, duplicates count: {data_dict[key].duplicated().sum()}')

book_data shape : (271360, 8), duplicates count: 0
rating_data shape : (1149780, 3), duplicates count: 0
users_data shape : (278858, 3), duplicates count: 0


In [5]:
# Info
for key in data_dict.keys():
  print(f'{key} info : {data_dict[key].info()}')
  print()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271360 entries, 0 to 271359
Data columns (total 8 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   ISBN                 271360 non-null  object
 1   Book-Title           271360 non-null  object
 2   Book-Author          271359 non-null  object
 3   Year-Of-Publication  271360 non-null  object
 4   Publisher            271358 non-null  object
 5   Image-URL-S          271360 non-null  object
 6   Image-URL-M          271360 non-null  object
 7   Image-URL-L          271357 non-null  object
dtypes: object(8)
memory usage: 16.6+ MB
book_data info : None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1149780 entries, 0 to 1149779
Data columns (total 3 columns):
 #   Column       Non-Null Count    Dtype 
---  ------       --------------    ----- 
 0   User-ID      1149780 non-null  int64 
 1   ISBN         1149780 non-null  object
 2   Book-Rating  1149780 non-null  int64 

In [6]:
# Describe
for key in data_dict.keys():
  display(data_dict[key].describe(include='all'))
  print()

ISBN      Book-Title      Book-Author  Year-Of-Publication  \
count       271360          271360           271359               271360   
unique      271360          242135           102023                  202   
top     0195153448  Selected Poems  Agatha Christie                 2002   
freq             1              27              632                13903   

        Publisher                                        Image-URL-S  \
count      271358                                             271360   
unique      16807                                             271044   
top     Harlequin  http://images.amazon.com/images/P/185326119X.0...   
freq         7535                                                  2   

                                              Image-URL-M  \
count                                              271360   
unique                                             271044   
top     http://images.amazon.com/images/P/185326119X.0...   
freq                                                    2   

                                              Image-URL-L  
count                                              271357  
unique                                             271041  
top     http://images.amazon.com/images/P/225307649X.0...  
freq                                                    2

User-ID        ISBN   Book-Rating
count   1.149780e+06     1149780  1.149780e+06
unique           NaN      340556           NaN
top              NaN  0971880107           NaN
freq             NaN        2502           NaN
mean    1.403864e+05         NaN  2.866950e+00
std     8.056228e+04         NaN  3.854184e+00
min     2.000000e+00         NaN  0.000000e+00
25%     7.034500e+04         NaN  0.000000e+00
50%     1.410100e+05         NaN  0.000000e+00
75%     2.110280e+05         NaN  7.000000e+00
max     2.788540e+05         NaN  1.000000e+01

User-ID                         Location            Age
count   278858.00000                           278858  168096.000000
unique           NaN                            57339            NaN
top              NaN  london, england, united kingdom            NaN
freq             NaN                             2506            NaN
mean    139429.50000                              NaN      34.751434
std      80499.51502                              NaN      14.428097
min          1.00000                              NaN       0.000000
25%      69715.25000                              NaN      24.000000
50%     139429.50000                              NaN      32.000000
75%     209143.75000                              NaN      44.000000
max     278858.00000                              NaN     244.000000

In [7]:
# Unique counts
for key in data_dict.keys():
  print('Unique counts for ',key)
  print(data_dict[key].nunique())
  print()

Unique counts for  book_data
ISBN                   271360
Book-Title             242135
Book-Author            102023
Year-Of-Publication       202
Publisher               16807
Image-URL-S            271044
Image-URL-M            271044
Image-URL-L            271041
dtype: int64

Unique counts for  rating_data
User-ID        105283
ISBN           340556
Book-Rating        11
dtype: int64

Unique counts for  users_data
User-ID     278858
Location     57339
Age            165
dtype: int64



In [8]:
# Unique values
print('Unique values for Year-Of-Publication')
print(book_data['Year-Of-Publication'].unique())

Unique values for Year-Of-Publication
[2002 2001 1991 1999 2000 1993 1996 1988 2004 1998 1994 2003 1997 1983
 1979 1995 1982 1985 1992 1986 1978 1980 1952 1987 1990 1981 1989 1984 0
 1968 1961 1958 1974 1976 1971 1977 1975 1965 1941 1970 1962 1973 1972
 1960 1966 1920 1956 1959 1953 1951 1942 1963 1964 1969 1954 1950 1967
 2005 1957 1940 1937 1955 1946 1936 1930 2011 1925 1948 1943 1947 1945
 1923 2020 1939 1926 1938 2030 1911 1904 1949 1932 1928 1929 1927 1931
 1914 2050 1934 1910 1933 1902 1924 1921 1900 2038 2026 1944 1917 1901
 2010 1908 1906 1935 1806 2021 '2000' '1995' '1999' '2004' '2003' '1990'
 '1994' '1986' '1989' '2002' '1981' '1993' '1983' '1982' '1976' '1991'
 '1977' '1998' '1992' '1996' '0' '1997' '2001' '1974' '1968' '1987' '1984'
 '1988' '1963' '1956' '1970' '1985' '1978' '1973' '1980' '1979' '1975'
 '1969' '1961' '1965' '1939' '1958' '1950' '1953' '1966' '1971' '1959'
 '1972' '1955' '1957' '1945' '1960' '1967' '1932' '1924' '1964' '2012'
 '1911' '1927' '1948' '1962' '2

In [9]:
# Unique values
print('Unique values for Book-Rating')
print(rating_data['Book-Rating'].unique())

Unique values for Book-Rating
[ 0  5  3  6  8  7 10  9  4  1  2]


In [10]:
# Unique values
print('Unique values for Age')
print(users_data['Age'].unique())

Unique values for Age
[ nan  18.  17.  61.  26.  14.  25.  19.  46.  55.  32.  24.  20.  34.
  23.  51.  31.  21.  44.  30.  57.  43.  37.  41.  54.  42.  50.  39.
  53.  47.  36.  28.  35.  13.  58.  49.  38.  45.  62.  63.  27.  33.
  29.  66.  40.  15.  60.   0.  79.  22.  16.  65.  59.  48.  72.  56.
  67.   1.  80.  52.  69.  71.  73.  78.   9.  64. 103. 104.  12.  74.
  75. 231.   3.  76.  83.  68. 119.  11.  77.   2.  70.  93.   8.   7.
   4.  81. 114. 230. 239.  10.   5. 148. 151.   6. 101. 201.  96.  84.
  82.  90. 123. 244. 133.  91. 128.  94.  85. 141. 110.  97. 219.  86.
 124.  92. 175. 172. 209. 212. 237.  87. 162. 100. 156. 136.  95.  89.
 106.  99. 108. 210.  88. 199. 147. 168. 132. 159. 186. 152. 102. 116.
 200. 115. 226. 137. 207. 229. 138. 109. 105. 228. 183. 204.  98. 223.
 113. 208. 107. 157. 111. 146. 118. 220. 143. 140. 189. 127.]


In [11]:
# Null/Missing Details
for key in data_dict.keys():
  print(f'Null count for {key} : ')
  print(data_dict[key].isna().sum())
  print()

Null count for book_data : 
ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

Null count for rating_data : 
User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

Null count for users_data : 
User-ID          0
Location         0
Age         110762
dtype: int64



### Data Details

1. I got three datasets, Books.csv, Ratings.csv, Users.csv
2. Shape and Duplicate count:
	* book_data shape : (271360, 8), duplicates count: 0
	* rating_data shape : (1149780, 3), duplicates count: 0
	* users_data shape : (278858, 3), duplicates count: 0
3. Books.csv :
	* Null count: 1 in Book Author,  2 in Publisher, 3 in Image-URL-L
	* Unique count :
	* ISBN : 271360
	* Book-Title : 242135
	* Book-Author : 102023
	* Year-Of-Publication : 202
	* Publisher : 16807
	* Image-URL-S : 271044
	* Image-URL-M : 271044
	* Image-URL-L : 271041
4. Rating.csv :
	* Unique count :
	* User-ID : 105283
	* ISBN : 340556
	* Book-Rating : 11
5. Users.csv :
	* Null count : 110762 in Age
	* Unique count :
	* User-ID : 278858
	* Location :57339
	* Age : 165

### EDA